pkg/geo/geogfn/segmentize.go

```diff
@@ -0,0 +1,161 @@
+				return nil, err
+			}
+			err = segPolygon.Push(l.(*geom.LinearRing))
+			if err != nil {
+				return nil, err
+			}
+		}
+		return segPolygon, nil
+	case *geom.MultiPolygon:
+		segMultiPolygon := geom.NewMultiPolygon(geom.XY).SetSRID(geometry.SRID())
+		for polygonIdx := 0; polygonIdx < geometry.NumPolygons(); polygonIdx++ {
+			p, err := segmentizeGeom(geometry.Polygon(polygonIdx), segmentMaxLength)
+			if err != nil {
+				return nil, err
+			}
+			err = segMultiPolygon.Push(p.(*geom.Polygon))
+			if err != nil {
+				return nil, err
+			}
+		}
+		return segMultiPolygon, nil
+	case *geom.GeometryCollection:
+		segGeomCollection := geom.NewGeometryCollection().SetSRID(geometry.SRID())
+		for geoIdx := 0; geoIdx < geometry.NumGeoms(); geoIdx++ {
+			g, err := segmentizeGeom(geometry.Geom(geoIdx), segmentMaxLength)
+			if err != nil {
+				return nil, err
+			}
+			err = segGeomCollection.Push(g)
+			if err != nil {
+				return nil, err
+			}
+		}
+		return segGeomCollection, nil
+	}
+	return nil, errors.Newf("unknown type: %T", geometry)
+}
+
+// segmentizeCoords inserts multiple points between given two-coordinate and
+// return resultant points as []float64. Such that distance between any two
+// points is less than given maximum segment's length, the total number of
+// segments is the power of 2, and all the segments are of the same length.
+// NOTE: List of points does not consist of end point.
+func segmentizeCoords(a geom.Coord, b geom.Coord, maxSegmentLength float64) []float64 {
+	// Converted geom.Coord into s2.Point so we can segmentize the coordinates.
+	point1 := s2.PointFromLatLng(s2.LatLngFromDegrees(a.Y(), a.X()))
+	point2 := s2.PointFromLatLng(s2.LatLngFromDegrees(b.Y(), b.X()))
+
+	var segmentizedPoints []s2.Point
+	distanceBetweenPoints := s2.ChordAngleBetweenPoints(point1, point2).Angle().Radians()
+	if maxSegmentLength <= distanceBetweenPoints {
+		// This calculation is to determine the total number of segment between given
+		// 2 coordinates. For that fraction by segment must be less than or equal to
+		// the fraction of max segment length to distance between point, since the
+		// total number of segment must be power of 2 therefore we can write as
+		// 1 / (2^n)[numberOfSegmentToCreate] <= segmentMaxLength / distanceBetweenPoints < 1 / (2^(n-1))
+		// (2^n)[numberOfSegmentToCreate] >= distanceBetweenPoints / segmentMaxLength > 2^(n-1)
+		// therefore n = ceil(log2(segmentMaxLength/distanceBetweenPoints)). Hence
+		// numberOfSegmentToCreate = 2^(ceil(log2(segmentMaxLength/distanceBetweenPoints))).
+		numberOfSegmentToCreate := int(math.Pow(2, math.Ceil(math.Log2(distanceBetweenPoints/maxSegmentLength))))
+		for pointInserted := 1; pointInserted < numberOfSegmentToCreate; pointInserted++ {
+			newPoint := s2.Interpolate(float64(pointInserted)/float64(numberOfSegmentToCreate), point1, point2)
+			segmentizedPoints = append(segmentizedPoints, newPoint)
+		}
+	}
+	allSegmentizedCoordinates := a.Clone()
```

Comment:
seems like we're iterating over the array twice, and allocate two arrays twice. can we restructure this to look like:

```
segmentizedCoordinates := []float64{}
segmentizedCoordinates = append(segmentizedCoordinates, a.Clone())
if maxSegmentLength <= distanceBetweenPoints {
   // ....
   for pointInserted := 1; pointInserted < numberOfSegmentToCreate; pointInserted++ {
      nextPoint := s2.Interpolate(....)
      segmentizedCoordinates = append(segmentizedCoordinates, nextPoint.Lng.Degrees(), latlng.Lat.Degrees())
  }
}
return segmentizedCoordinates
```